Async endeavour: Use promises

mkg20001 opened the pull request:
I've implemented secio in an async way

~~For now only the main methods have been done, as I'm unsure whether or not the approach taken by me fits into the libp2p ecosystem~~ I'm refactoring the whole thing, yet it would be great if somebody could review it during the refactor, as I'm unsure whether or not the approach taken by me fits into the libp2p ecosystem

Basically: The callback(s) have been completely replace by a new variable `encryptedConnection.awaitConnected` that can be `await`ed to see if any errors occured

mkg20001 commented:
Seems to mysteriously hang at reading proposal, sometimes
![screenshot](https://user-images.githubusercontent.com/7735145/61460883-ef22a580-a96f-11e9-8f7b-d36d2f401ca0.png)

mkg20001 commented:
Ref https://github.com/ipfs/js-ipfs/issues/1670

mkg20001 commented:
Seems like it sometimes sets `remoteId` to undefined and then there's uncaught crash in src/index.js where it tries to create a peerinfo from the peerid that is undefined

mkg20001 commented:
With the [new interface-connection](https://github.com/libp2p/interface-connection/pull/29) being finished soon, should I first just port secio to async/await or already implement the async iterators?

jacobheun commented:
@mkg20001 async iterators. I'm working on the internal refactor for the switch around the crypto upgrade.  I am implementing the plaintext 2.0.0 spec in async iterators as part of that to make sure the upgrade process works smoothly, as well as sorting out any bugs with the current async updates.

mkg20001 commented:
@jacobheun Could you give me some guidance on how to implement them?

Btw, it seems like I can't push to this branch

jacobheun commented:
@mkg20001 yes, once I get the initial dial flow working and any bugs worked out I am aiming to create a contributor guide for the async iterators to help us migrate the remaining modules. Hoping to get started on that in the next couple days.

mkg20001 commented on .aegir.js:
```diff
@@ -4,6 +4,7 @@ const multiaddr = require('multiaddr')
 const pull = require('pull-stream')
 const WS = require('libp2p-websockets')
 const PeerId = require('peer-id')
+const prom = (fnc) => new Promise((resolve, reject) => fnc((err, res) => err ? reject(err) : resolve(res)))
```
This turns callbacks into promises

daviddias replied:
Why not use https://www.npmjs.com/package/promisify ? (I mean, if this one while works perfectly for all cases, then great)

mkg20001 replied:
It covers most of the use cases and doesn't require an extra module. Also it has a more convenient syntax: Instead of having to first promisify a module and then use it, I can simply do `prom(cb => anything(anyArgument, cb))` and await that.

I could toss it into a module, but I found that overkill since it's pretty easy to remember after using it for a while

jacobheun commented:
@mkg20001 I opened a PR for a crypto interface for the async work, https://github.com/libp2p/js-interfaces/pull/2, which also includes a crypto mock in the test suite for those tests which shows how the crypto streams might get wrapped internally. I'm going to be building plaintext/2.0.0 off of that. I also added an example and some resources to the streaming iterables guide in js-libp2p, https://github.com/libp2p/js-libp2p/pull/466.

mkg20001 commented:
I'm currently working on a module that will make working with protocol-buffers based RPC easier (just a tiny shim around reading length-prefixed data). Once that is ready, I'll continue working on this here.

daviddias commented on src/support.js:
```diff
@@ -121,3 +120,7 @@ exports.read = function read (reader) {
   })
 }
+
+exports.wrap = (duplex) => {
```
What's this for?

mkg20001 replied:
:sweat_smile: some leftovers.

Currently pushed some broken code, just to iterate a bit how to do the streaming.

Unsure where I should wrap the stream: Before passing it to handshake and then using that object (as I did now) or using the support lib and a handshake stream.

mkg20001 commented:
Blocked by https://github.com/alanshaw/it-length-prefixed/issues/5

mkg20001 commented:
(/me is currently trying to make sense of it-handshake: https://github.com/jacobheun/it-handshake/issues/1)

mkg20001 commented:
Stuff that's still broken:
 - [ ] it-pb-rpc .read() seems to return null for some reason, breaking the 3. finish part https://github.com/mkg20001/it-pb-rpc/issues/1
 - [ ] it-length-prefixed isn't supporting the "fixed" prefix type https://github.com/alanshaw/it-length-prefixed/issues/5
 - [ ] it-length-prefixed PR for decodeFromReader is still unmerged https://github.com/alanshaw/it-length-prefixed/pull/3
   - [ ] also update it-pb-rpc

mkg20001 commented:
Things that need to be considered:
 - [ ] the new it-handshake doesn't have a timeout option. do we need it? or are timeouts handled in the switch?

jacobheun commented:
> the new it-handshake doesn't have a timeout option. do we need it? or are timeouts handled in the switch?

Support could be added there, but we won't need it for libp2p. Since we can now pass an AbortController if the connection process exceeds the libp2p timeout we can abort it at the transport level and let that bubble up.

mkg20001 commented:
So it "works", but beware: It is currently using varint length-prefixes instead of fixed-size (32bit int) ones, so it's incompatible.
Once https://github.com/alanshaw/it-length-prefixed/issues/5 is added this should be fixed

jacobheun commented on src/index.js:
```diff
@@ -1,58 +1,34 @@
 
-      conn.getPeerInfo((err, peerInfo) => {
-        encryptedConnection.setInnerConn(new Connection(state.secure, conn))
+  const timeout = 60 * 1000 * 5
```
Wow, this is a 5 minute timeout?! I know it's unchanged from master, but this is going to end up exceeding the timeout of the entire connection upgrade, now that the libp2p dialer takes the entire connection upgrade into account. We can probably leave it for now, but unless someone specifically increases their dial timeouts this won't ever get triggered as the connection will get aborted long before.

jacobheun replied:
Actually, let's just take this out. The default timeout of `State` is 60s, we should probably just use that.

mkg20001 replied:
As mentioned in the PR, this is no longer getting used. We can remove it entierly

jacobheun commented on src/support.js:
```diff
@@ -39,97 +36,64 @@ exports.theBest = (order, p1, p2) => {
-    callback()
-  })
+  target.mac = await macAndCipher[0] // ..but get awaited after each other
+  target.cipher = await macAndCipher[1]
```
Any reason to not just do the following?

```js
[target.mac, target.cipher] = await Promise.all([
  makeMac(target.hashT, target.keys.macKey),
  makeCipher(target.cipherT, target.keys.iv, target.keys.cipherKey)
])
```

mkg20001 replied:
Opinions. Which one is more readable in your opinion?

jacobheun replied:
The latter, imo, as I think it more clearly shows the parallelism, but this doesn't need to change. I think the second comment `// ..but get awaited after each other` might have been the thing that made me look at this twice.

jacobheun commented on test.old/browser.js:
```diff
@@ -10,7 +10,6 @@ const pull = require('pull-stream')
 const pullGoodbye = require('pull-goodbye')
```
Right now none of the tests here are running, just the `interface` test suite. We should just remove all of these and .aegir.js, and add tests for any utility / secio specific methods. The interface test suite is only verifying the high level api, not that the crypto exchange is correct, we should add tests for this.

jacobheun replied:
For example, it would be great to see a failing test showing the issue with temporarily using varint instead of fixed size. While it's working, we know this does not fit the spec. I'd expect to see the interface tests pass and the local tests failing there.

mkg20001 replied:
Don't know how to test that without any lib supporting that. We'd need an interop test or sth.

jacobheun replied:
It will likely be easier once `it-length-prefixed` supports it, but we should be able to validate that the payload is what's being expected. Mock data could be passed to the `finish` method along with a spy on the `it-pb-rpc` wrapper to validate that when data is sent over one duplex pair half to the other, that the encrypted payload is as expected. Interop isn't really needed here, we just need to validate we're sending the right thing.

mkg20001 commented on benchmarks/send.js:
```diff
@@ -1,73 +1,61 @@
-  const aToB = secio.encrypt(peerA, new Connection(p[0]), peerB, ifErr)
-  const bToA = secio.encrypt(peerB, new Connection(p[1]), peerA, ifErr)
+  const aToB = await secio.secureInbound(peerA, p[0], peerB)
+  const bToA = await secio.secureOutbound(peerB, p[0], peerA)
```
Indeed. Wondering why it worked at all...

mkg20001 commented on src/etm.js:
```diff
@@ -1,81 +1,41 @@
+    for await (const chunk of source) {
+      const data = await cipher.encrypt(chunk)
+      const digest = await mac.digest(data)
+      yield new BufferList().append(data).append(digest)
```
I saw it in one of @alanshaw's modules that way :sweat_smile:

mkg20001 commented on README.md:
```diff
@@ -46,17 +46,18 @@ const secio = require('libp2p-secio')
 The current `secio` tag, usable in `multistream`.
 
-### `const encryptedConnection = secio.encrypt(localPeerId, plainTextConnection [, remotePeerId] [, callback])`
+### `const encryptedConnection = secio.encrypt(localPeerId, plainTextConnection [, remotePeerId])`
```
To `js-interfaces` README?

jacobheun replied:
Yes, more specifically the crypto api section at https://github.com/libp2p/js-interfaces/tree/master/src/crypto#api

jacobheun commented:
@mkg20001 `it-length-prefixed` 3.0.0 is out now, https://github.com/alanshaw/it-length-prefixed. It includes `lp.decode.fromReader` (was decodeFromReader in my branch), and support for the custom length coder which is included there for 32intBE.

mkg20001 commented:
@jacobheun I've updated it-pb-rpc and added the LP encoder/decoder. I'll fix the rest of the mentioned comments and then this should be ready! :tada:

alanshaw commented on src/handshake/finish.js:
```diff
@@ -12,9 +12,16 @@ const ensureBuffer = require('it-buffer')
+  target.writeInt32BE(value, offset)
+  return target
+}
+lengthEncoder.bytes = 4 // Always because fixed length
```
This can just be imported from `it-length-prefixed`

jacobheun commented:
There was an issue with benchmarks that was causing the performance to appear to be much higher than it actually is. `benchmark` doesn't support promises so the deferred still needs to be called. I fixed the issue and the results are actually similar to benchmarks against master, there's not a significant difference.

I am looking into an issue with compatibility against the current version of libp2p. I believe there are a couple places we are still using varint that is causing the issue. I'm looking to fix that today so we can get this merged.

alanshaw commented on package.json:
```diff
@@ -25,31 +25,33 @@
   "license": "MIT",
   "dependencies": {
-    "async": "^2.6.2",
+    "async": "^3.1.0",
```
Still needed? If so, can we please remain at v2?

alanshaw commented on package.json:
```diff
@@ -25,31 +25,33 @@
+    "libp2p-interfaces": "~0.1.3",
+    "multiaddr": "^7.2.1",
+    "multihashing-async": "~0.8.0",
     "once": "^1.4.0",
```
Still needed?

jacobheun commented:
I created a PR in https://github.com/mkg20001/it-pb-rpc/pull/2 to add support for passing the fixed encoder. The last issue with compatibility I am working on is that this branch is currently just sending the final handshake of nonces over the unencrypted connection. The nonces have to be sent over the newly constructed crypto box streams so that the encryption can be validated.

mkg20001 commented:
> The nonces have to be sent over the newly constructed crypto box streams so that the encryption can be validated.

Aren't they currently?

jacobheun commented:
> Aren't they currently?

No, the write and read in `finish` are being done before the box/unbox pipe is created. It's easiest to see it in the network traffic, only the 16 bytes of the nonce are exchanged.

I pushed up https://github.com/libp2p/js-libp2p-secio/pull/108/commits/840c217dbbd3baa491d2eec0b0349b708e1ab4c1 which fixes it. I am working on adding a test that will verify the exchange is compliant with the spec.

jacobheun commented:
I've added the tests for inbound and outbound (which are the same since it's symmetric) encryption so we can verify the upgrade process. I also updated the dependency list and added node 12 to CI.

Here are the most recent benchmarks:
**Async/await**
```log
create peers for test x 653 ops/sec ±29.95% (68 runs sampled)
establish an encrypted channel x 207 ops/sec ±1.65% (77 runs sampled)
send plaintext 10 x 262144 bytes x 5,226 ops/sec ±2.34% (71 runs sampled)
send encrypted 10 x 262144 bytes x 41.79 ops/sec ±1.18% (65 runs sampled)
send plaintext 100 x 262144 bytes x 547 ops/sec ±3.42% (26 runs sampled)
send encrypted 100 x 262144 bytes x 5.21 ops/sec ±0.93% (29 runs sampled)
send plaintext 1000 x 262144 bytes x 52.61 ops/sec ±2.70% (59 runs sampled)
send encrypted 1000 x 262144 bytes x 0.55 ops/sec ±0.50% (7 runs sampled)
```

**master**
```log
create peers for test x 646 ops/sec ±28.13% (70 runs sampled)
establish an encrypted channel x 192 ops/sec ±1.52% (78 runs sampled)
send plaintext 10 x 262144 bytes x 4,135 ops/sec ±2.82% (62 runs sampled)
send encrypted 10 x 262144 bytes x 38.24 ops/sec ±1.09% (62 runs sampled)
send plaintext 100 x 262144 bytes x 521 ops/sec ±2.19% (70 runs sampled)
send encrypted 100 x 262144 bytes x 4.91 ops/sec ±0.78% (28 runs sampled)
send plaintext 1000 x 262144 bytes x 53.99 ops/sec ±3.25% (61 runs sampled)
send encrypted 1000 x 262144 bytes x 0.51 ops/sec ±1.49% (7 runs sampled)
```

I have also run a test with an echo protocol of `libp2p@latest` and `libp2p@refactor/async` over tcp with mplex and secio. Dials and echos are working both ways, this should be good to go!

jacobheun commented:
We resolved npm perms issues, the `beta` version 0.12.x is now on npm.  @mkg20001 thank you for all your work on this!
